cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -1514,14 +1514,19 @@ class AsofJoinNode : public ExecNode {
   }
 
   Status StartProducing() override {
-    ARROW_ASSIGN_OR_RAISE(process_task_, plan_->query_context()->BeginExternalTask(
-                                             "AsofJoinNode::ProcessThread"));
-    if (!process_task_.is_valid()) {
-      // Plan has already aborted.  Do not start process thread
+    #ifndef ARROW_ENABLE_THREADING
+      return Status::NotImplemented("ASOF join requires threading enabled");
+    #else
```

Comment:
Can we use the following approach?

```cpp
Status StartProducing() override {
#ifndef ARROW_ENABLE_THREADING
  return Status::NotImplemented("ASOF join requires threading enabled");
#endif
  ARROW_ASSIGN_OR_RAISE(process_task_, plan_->query_context()->BeginExternalTask(
  // use existing code as-is.
```

Putting existing code to `#else ... #endif` reduces readability.